Book.java

```diff
@@ -0,0 +1,315 @@
+package accounting;
+
+import agents.Agent;
+import actions.Action;
+import contracts.Asset;
+import contracts.Contract;
+import contracts.Loan;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+
+/**
+ * This is the main class implementing double entry accounting. All public operations provided by this class
+ * are performed as a double entry operation, i.e. a pair of (dr, cr) operations.
+ *
+ * A Book contains a set of accounts, and is the interface between an agent and its accounts. Agents cannot
+ * directly interact with accounts other than via a Book.
+ *
+ * At the moment, a Book contains an account for each type of contract, plus an equity account and a cash account.
+ *
+ * A simple economic agent will usually have a single Book, whereas complex firms and banks can have several books
+ * (as in branch banking for example).
+ *
+ * @author rafa
+ */
+public class Book implements BookAPI {
+
+    public Book(Agent owner) {
+        // A Book is a list of accounts
+        accounts = new HashSet<>();
+
+        // Subsets of the list of accounts (for quicker searching)
+        assetAccounts = new HashSet<>();
+        liabilityAccounts = new HashSet<>();
+        equityAccounts = new HashSet<>();
+
```

Comment:
Is it strictly necessary to have different hash sets holding the same account objects? Although (I think) Java stores values by reference, I fear there might be situations where, for some reason, the account object gets copied and thus ends up being 'forked'.  

Would it be better to use just one single data structure (e.g. `assetAccountsMap` described below) and use clever access functions to filter for the type of content that you're interested in? For example, to return all accounts from `assetAccountsMap`, in a lazy way (i.e. as a stream), you could define the following function (sorry, haven't checked if it compiles):

```
Stream<Object> getAccounts() {
  return assetAccountsMap.entrySet().stream.map(Map.Entry::getValue);
}
```

To return all accounts of a particular type, you could define the following function:

```
Stream<Object> getAccountsOfType(AccountType t) {
  return assetAccountsMap.entrySet().stream().filter(entry -> entry.getKey() == t).map(Map.Entry::getValue);
}
```

In order to turn the streams into lists, you could also define a general function:

```
List<?> toList(Stream<Object> s) {
  return s.collect(Collectors.toList());
}
```

As I said, I haven't checked whether the snippets compile, so please take them with a pinch of salt :)